### PR TITLE
Local DNS system for connecting apps together via host machine

### PR DESCRIPTION
**Pull request by thenonsequitur**
## Problem

There is currently no convenient system for linking local dockerized apps together.

We have a rudimentary system for linking some apps together. For example, a local instance of chorus can connect to a local instance of hermano by linking the app default networks together and passing in a hostname ENV var. This is sub-optimal for two main reasons:

1. It requires several manual steps (not very intuitive steps either).

2. Chorus in particular requires real local hostnames (e.g. sbnation--com.local.sbndev.net) to work correctly, so anything using chorus for oauth or api (e.g. anthem) would have to use a manual hack to deal with this issue.

## Solution

This PR adds a new base image that runs a dnsmasq server (on alpine!). The configuration for the dnsmasq server here is very minimal and basically just says "If the hostname is *.local.sbndev.net, resolve to the host machine ip address, otherwise serve dns as regular".

Apps will be able to add this dns container to their docker-compose files, and if the app service (or any other service) is told to use this dns server, then they can reference chorus local hostnames, which in turn will call the host. For example, anthem can reference `sbnation--com.local.sbndev.net:3000` and, assuming that chorus is running on the host machine served at port 3000, it will just work, with no extra steps or network connections required.

Currently, this is only set up for chorus local hostnames, but any other apps that are hostname-aware can extend this functionality.

And for apps that don't care about the hostname, you can still use the plain `local.sbndev.net` address to reference the host machine. For example, if hermano is running on port 5000, anthem can connect to hermano at `local.sbndev.net:5000`, and again it will just work with no extra steps required.

## How it works

The local dns server container (which will be tagged `voxmedia/local-dns`) is just a dnsmasq server that reads all configuration files from /etc/dnsmaq. It has a built-in entrypoint script that determines the host ip address (in a host-os-agnostic way) and emits a dnsmasq config file to map `*.local.sbndev.net` to this ip address.

Apps will be able to include this dns service with a simple docker-compose directive:

```
services:
  dns:
    image: voxmedia/local-dns:latest
```

But we still need a way to tell the app service (or any other service that needs access to this local dns) to use this dns server for resolution. Docker-compose has a [simple directive](https://docs.docker.com/compose/compose-file/#dns) to specify alternate dns servers, but unfortunately this only works with static addresses. This won't work here because the dns service, like any other compose service, is dynamically assigned an ip address when the container is started. So there is no way to know the ip address ahead of time.

To get around this limitation, we use a hack. There is a new entrypoint script in the ruby base image, `use_dns_server.sh` that simply finds the ip address of the dns server and then replaces `/etc/resolv.conf` with a directive to use this ip address as the nameserver. This is pretty ugly, but there doesn't seem to be a cleaner way to do this.

The app service can then be updated to reference the dns service as a dependency, e.g.:

```
services:
  dns:
    image: voxmedia/local-dns:latest
  app:
    ...
    depends_on:
      ...
      - dns
```

And then the following script added to the entrypoint:

```
/opt/entrypoint/use_dns_server.sh dns
```

Then the app service can freely send requests to `*.local.sbndev.net` addresses and these requests will go to the host (and all other dns will resolve as it did previously). This isn't limited to the app service -- any service that needs this local dns can be set up the same way.

## A note about versions

We are moving the base images to semantic versioning. This new release of the base image will be `1.0.0` and the new local-dns service is also starting out at version `1.0.0`.

## Tests

This was all verified locally. Using this system, I successfully connected anthem, chorus, and hermano together with no extra steps, by simply setting the appropriate `local.sbndev.net` addresses in the local.yml config file.

## Launch plan

After merging, build and release new images for both the ruby base image and the new local-dns image.

Note that we are still using public repos on docker hub, so this will technically be public, with the only security being obscurity. While it's not ideal that a potential attacker might see that we are using `local.sbndev.net`, there's nothing particularly unsafe about this being exposed...there's no way anyone  could use this information as part of an exploit. And this will all be a moot point if we move to private docker repos in the future.

After it's merged, will apply it to anthem and confirm it works as-is, then update wiki docs on how to use it.

### CODE REVIEW

File: local-dns/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+FROM alpine:latest
+
+RUN apk --no-cache add dnsmasq
```
Comment (cschiewek) [inline local-dns/Dockerfile]: we'll probably want `rm -rf /var/cache/apk/*` here to remove the apk cache to keep the image as small as possible.

Comment (cschiewek) [reply local-dns/Dockerfile]: nvm, I didn't see the no-cache flag cause I can't read

Comment (thenonsequitur) [reply local-dns/Dockerfile]: I thought that's what `--no-cache` does already. Is your understanding different?

**Comment by ebinmore** [comment]
@ckolderup I believe you just need to reference in your `Dockerfile`  a la https://github.com/voxmedia/sbn/blob/master/Dockerfile#L28 which calls the copy from the base image.

**Pull request merged**